synthesis/codon/codon.go

```diff
@@ -142,66 +248,46 @@ func Translate(sequence string, codonTable Table) (string, error) {
 	return aminoAcids.String(), nil
 }
 
-// Optimize takes an amino acid sequence and codonTable and returns an optimized codon sequence. Takes an optional random seed as last argument.
-func Optimize(aminoAcids string, codonTable Table, randomState ...int) (string, error) {
-	// Finding any given aminoAcid is dependent upon it being capitalized, so
-	// we do that here.
-	aminoAcids = strings.ToUpper(aminoAcids)
-
-	if codonTable.IsEmpty() {
-		return "", errEmptyCodonTable
-	}
-	if len(aminoAcids) == 0 {
-		return "", errEmptyAminoAcidString
-	}
-
-	// weightedRand library insisted setting seed like this. Not sure what environmental side effects exist.
-	if len(randomState) > 0 {
-		rand.Seed(int64(randomState[0]))
-	} else {
-		rand.Seed(time.Now().UTC().UnixNano())
-	}
-
-	var codons strings.Builder
-	codonChooser, err := codonTable.Chooser()
-	if err != nil {
-		return "", err
-	}
-
-	for _, aminoAcid := range aminoAcids {
-		chooser, ok := codonChooser[string(aminoAcid)]
-		if !ok {
-			return "", invalidAminoAcidError{aminoAcid}
-		}
-		codons.WriteString(chooser.Pick().(string))
-	}
-	return codons.String(), nil
-}
-
-// OptimizeTable weights each codon in a codon table according to input string codon frequency.
-// This function actually mutates the codonTable struct itself.
-func (table codonTable) OptimizeTable(sequence string) Table {
+// weightAminoAcids weights each codon in a codon table according to input string codon frequency, adding weight to
+// the given NCBI base codon table
+func weightAminoAcids(sequence string, aminoAcids []AminoAcid) []AminoAcid {
 	sequence = strings.ToUpper(sequence)
 	codonFrequencyMap := getCodonFrequency(sequence)
 
-	for aminoAcidIndex, aminoAcid := range table.AminoAcids {
+	for aminoAcidIndex, aminoAcid := range aminoAcids {
 		// apply weights to codonTable
 		for codonIndex, codon := range aminoAcid.Codons {
-			table.AminoAcids[aminoAcidIndex].Codons[codonIndex].Weight = codonFrequencyMap[codon.Triplet]
```

Comment:
If I'm not mistaken this function updated the (global) default NCBI tables, it seems that this may have been intended but I assumed not so changed `getCodonTable()` to return a copy of the table rather than the hardcoded one.

Reply:
The global NCBI tables was intended so you only had to instantiate it once, but might be better to have it under a function. The NCBI tables really don't mutate quickly at all, so it'd be nicest to have them as constants, but that is kinda hard in Golang I think